lib/truncate.js

```diff
@@ -10,192 +10,222 @@ exports.error = truncError
-        }
-        break
-
-      case 'system':
-        switch (path[1]) {
-          case 'architecture':
-          case 'hostname':
-          case 'platform':
+      let max = opts.truncateStringsAt
+      switch (path[0]) {
+        case 'service':
+          switch (path[1]) {
+            case 'name':
+            case 'version':
+            case 'environment':
+              max = opts.truncateKeywordsAt
+              break
+
+            case 'agent':
+            case 'framework':
+            case 'language':
+            case 'runtime':
+              switch (path[2]) {
+                case 'name':
+                case 'version':
+                  max = opts.truncateKeywordsAt
+                  break
+              }
+              break
+          }
+          break
+
+        case 'process':
+          if (path[1] === 'title') {
             max = opts.truncateKeywordsAt
-            break
-        }
-        break
+          }
+          break
+
+        case 'system':
+          switch (path[1]) {
+            case 'architecture':
+            case 'hostname':
+            case 'platform':
+              max = opts.truncateKeywordsAt
+              break
+          }
+          break
+      }
+
+      return truncate(value, max)
     }
+  })
+}
 
-    return truncate(value, max)
-  }, true)
+function onObject (value, key, path, isNew) {
+  if (isHeader(path)) return value.toString()
+  return isNew ? {} : '[Circular]'
 }
 
 function truncTransaction (trans, opts) {
-  trans.name = truncate(String(trans.name), opts.truncateKeywordsAt)
-  trans.type = truncate(String(trans.type), opts.truncateKeywordsAt)
-  trans.result = truncate(String(trans.result), opts.truncateKeywordsAt)
+  const result = breadthFilter(trans, {
+    onObject,
+    onValue (value, key, path) {
+      if (typeof value !== 'string') {
+        if (isHeader(path)) return String(value)
 
-  breadthFilter(trans, (value, key, path) => {
-    if (typeof value !== 'string') {
-      return value
-    }
-
-    let max = opts.truncateStringsAt
-    switch (path[0]) {
-      case 'name':
-      case 'type':
-      case 'result':
         return value
+      }
 
-      case 'id':
-      case 'trace_id':
-      case 'parent_id':
-        max = opts.truncateKeywordsAt
-        break
+      let max = opts.truncateStringsAt
+      switch (path[0]) {
+        case 'name':
+        case 'type':
+        case 'result':
+          value = String(value)
+          max = opts.truncateKeywordsAt
+          break
+
+        case 'id':
+        case 'trace_id':
+        case 'parent_id':
+          max = opts.truncateKeywordsAt
+          break
 
-      case 'context':
-        max = contextLength(path, opts)
-        break
+        case 'context':
+          max = contextLength(path, opts)
+          break
+      }
+
+      return truncate(value, max)
     }
+  })
 
-    return truncate(value, max)
-  }, true)
+  return Object.assign({
+    name: 'undefined',
```

Comment:
What's the thought behind these `"undefined"` string values?

Reply:
This is reproducing the behaviour from before where `trans.name = truncate(String(trans.name), opts.truncateKeywordsAt)` would produce `"undefined"` if the name property was undefined because of the explicit string conversion. The `breadthFilter` will skip the field entirely, if not present as it can't operate on a field it's not aware of.

Reply:
Hm true. I'm wondering about the overhead though. Hopefully these properties will always be present. But lets go with it for now